### DISCUSSION TITLE
Add Optimize option to Arduino board configuration

### DISCUSSION THREAD

**Discussion by Prow7** [earlephilhower/arduino-pico#303] [Ideas]
It would be nice to have a compiler optimize option in the Arduino board configuration, like the Adafruit SAMD ItsyBitsy boards have:
Small (-Os)
Fast (-O2)
Faster (-O3)
Fastest (-Ofast)
Here be dragons (-Ofast -funroll-loops)

I did a bit of investigating to figure out where this is configured but didn't do so far as to make a formal change for a PR. Figured I'd ask here first.

Cheers.

**Comment by earlephilhower** [comment]
@Prow7 this would be easiest added to https://github.com/earlephilhower/arduino-pico/blob/master/tools/makeboards.py .

The caveat might be, though, that with SPI flash for ROM the higher level optimizations actually make things go slower.  `-O2` probably would be fine, but when you start unrolling loops and increasing code size, the instruction cache misses increase (and each miss is painfully slow).

A PR would be welcome.  Make a new menu in the python script to set a `build.optimize` variable, and use that in platform.txt instead of the fixed `-Os`.

**Comment by Ing-Dom** [comment]
no optimazation should also be an option, at least for debugging

**Comment by Prow7** [comment]
Great point about the higher levels actually making the code execute slower. My code is only a single example but -O3 appears to be the fastest setting. -Ofast actually executes a bit slower. Now I'm wondering if the -Ofast options should be included.
